### PR TITLE
Log transaction hook errors for better UX

### PR DESCRIPTION
**Pull request by cicdw**
I noticed that errors in both rollback and commit hooks are completely swallowed, which makes it really hard to debug when they occur.  This PR adds logging for those errors.

### CODE REVIEW

File: src/prefect/transactions.py [hunk 1]
```diff
@@ -58,6 +61,7 @@ class Transaction(ContextModel):
         default_factory=list
     )
     overwrite: bool = False
+    logger: Optional[PrefectLogAdapter] = None
```
Comment (desertaxle) [inline src/prefect/transactions.py:64]: I assume we could handle any type of logger here
```suggestion
    logger: Optional[Logger] = None
```

**Review by desertaxle — APPROVED** [review]
LGTM! Left a couple of suggestions, but they're non-blocking

File: src/prefect/transactions.py (hunk 1 above)
Comment (cicdw) [reply src/prefect/transactions.py:64]: I tried that type hint, but it failed! Within the engine we use `task_run_logger` which apparently outputs a ... different type of object?

File: src/prefect/transactions.py [hunk 2]
```diff
@@ -288,7 +305,16 @@ def transaction(
             result_factory=new_factory,
         )
 
+    try:
+        logger = logger or get_run_logger()
+    except MissingContextError:
+        logger = None
```
[the remark this answers is not in the document]
Comment (cicdw) [reply src/prefect/transactions.py:311]: yea good call, will update along with an import

**Review by chrisguidry — APPROVED** [review]
Nice!

File: src/prefect/transactions.py (hunk 1 above)
Comment (desertaxle) [reply src/prefect/transactions.py:64]: That's whack. I guess this type hint would be the most technically correct then:
```suggestion
    logger: Union[logging.Logger, logging.LoggerAdapter, None] = None
```

Comment (cicdw) [reply src/prefect/transactions.py]: Begrudgingly updated lol

**Pull request merged**